protocol-select/README.md

```diff
@@ -0,0 +1,511 @@
+a nested negotiation - on a stream.
+
+#### Initiator
+
+The initiator of a new stream is the first endpoint to send a message. We
+differentiate in the following two scenarios.
+
+1. **Optimistic Protocol Negotiation**: The endpoint knows exactly which
+   protocol it wants to use. It then only sends this protocol. It MAY start
+   sending application data right after the _Protocol Select_ protobuf message.
+   Since it has not received confirmation from the remote peer for the protocol,
+   any such data might be lost in such case.
+
+2. **Multi Protocol Negotiation**: The endpoint wants to use any of a set of
+   protocols, and lets the remote peer decide which one. It then sends the list
+   of protocols. It MUST wait for the peer's protocol choice before sending
+   application data.
+
+An initiator MUST treat the receipt of an empty list of protocols in response to
+its list of protocols as a negotiation failure and thus a stream error.
+
+#### Listener
+
+The listening endpoint replies to a list of protocols from the initiator by
+either:
+
+- Sending back a single entry list with the protocol it would like to speak.
+
+- Rejecting all proposed protocols by replying with an empty list of protocols.
+
+## Transitioning from Multistream Select
+
+Protocol Select is not compatible with [Multistream Select] both in its
+semantics as well as on the wire. Live libp2p-based networks, currently using
+[Multistream Select], will need to follow the multiphased roll-out strategy
+detailed below to guarantee a smooth transition.
+
+### Multiphase Rollout
+
+#### Phase 1
+
+In the first phase of the transition from [Multistream Select] to Protocol
+Select, nodes in the network are upgraded to support both [Multistream Select]
+and Protocol Select when accepting inbound connections, i.e. when acting as a
+listener. Differentiating the two protocols as a listener is detailed in the
+[Heuristics](#heuristics) section below. Nodes, when dialing, MUST NOT yet use
+Protocol Select, but instead continue to use [Multistream Select].
+
+Once a large enogh fraction of the network has upgraded, one can transition to
+phase 2.
+
+#### Phase 2
+
+With a large enough fraction of the network supporting Protocol Select on
+inbound connections, nodes MAY start using Protocol Select on outbound
+connections.
+
+After a large enough fraction of the network has upgraded, i.e. uses Protocol
+Select for outbound connections, one can transition to phase 3.
+
+#### Phase 3
+
+Given that a large enough fraction of the network uses Protocol Select for both
+out- and inbound connections, nodes can drop support for [Multistream Select]
+concluding the transition.
+
+### Additional Rollout Mechanisms
+
+When attempting to upgrade an outbound connection with Protocol Select to a node
+that does not yet support Protocol Select, the connection attempt will fail:
+
+* TCP: when the cryptographic handshake is started
+
+* QUIC: when the first stream is opened
```

Comment:
Do we really need to close a connection? I would say in case of QUIC one can open another substream using Multistream Select.

Reply:
In theory that would work, but it might be quite involved implementation-wise.